envelope.go

```diff
@@ -25,286 +25,177 @@ import (
-		if err := r.compressionPool.Decompress(decompressed, data, int64(r.readMaxBytes)); err != nil {
-			return err
-		}
-		data = decompressed
-	}
-
-	if env.Flags != 0 && env.Flags != flagEnvelopeCompressed {
-		// Drain the rest of the stream to ensure there is no extra data.
-		if n, err := discard(r.reader); err != nil {
-			return errorf(CodeInternal, "corrupt response: I/O error after end-stream message: %w", err)
-		} else if n > 0 {
-			return errorf(CodeInternal, "corrupt response: %d extra bytes after end of stream", n)
+func readAll(dst *bytes.Buffer, src io.Reader, readMaxBytes int) *Error {
+	var totalN int64
+	for {
+		readN, err := read(dst, src)
+		totalN += int64(readN)
+		if readMaxBytes > 0 && totalN > int64(readMaxBytes) {
+			discardN, err := discard(src)
+			if err != nil {
+				return errorf(CodeResourceExhausted,
+					"message is larger than configured max %d - unable to determine message size: %w",
+					readMaxBytes, err)
+			}
+			return errorf(CodeResourceExhausted,
+				"message size %d is larger than configured max %d",
+				totalN+discardN, readMaxBytes)
 		}
-		// One of the protocol-specific flags are set, so this is the end of the
-		// stream. Save the message for protocol-specific code to process and
-		// return a sentinel error. Since we've deferred functions to return env's
-		// underlying buffer to a pool, we need to keep a copy.
-		copiedData := make([]byte, data.Len())
-		copy(copiedData, data.Bytes())
-		r.last = envelope{
-			Data:  bytes.NewBuffer(copiedData),
-			Flags: env.Flags,
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+			if writeErr, ok := asError(err); ok {
+				return writeErr
+			}
+			if readMaxBytesErr := asMaxBytesError(err, "read first %d bytes of message", totalN); readMaxBytesErr != nil {
+				return readMaxBytesErr
+			}
+			return errorf(CodeUnknown, "read: %w", err)
 		}
-		return errSpecialEnvelope
-	}
-
-	if err := r.codec.Unmarshal(data.Bytes(), message); err != nil {
-		return errorf(CodeInvalidArgument, "unmarshal into %T: %w", message, err)
 	}
-	return nil
 }
 
-func (r *envelopeReader) Read(env *envelope) *Error {
-	prefixes := [5]byte{}
-	prefixBytesRead, err := r.reader.Read(prefixes[:])
-
-	switch {
-	case (err == nil || errors.Is(err, io.EOF)) &&
-		prefixBytesRead == 5 &&
-		isSizeZeroPrefix(prefixes):
-		// Successfully read prefix and expect no additional data.
-		env.Flags = prefixes[0]
-		return nil
-	case err != nil && errors.Is(err, io.EOF) && prefixBytesRead == 0:
-		// The stream ended cleanly. That's expected, but we need to propagate them
-		// to the user so that they know that the stream has ended. We shouldn't
-		// add any alarming text about protocol errors, though.
-		return NewError(CodeUnknown, err)
-	case err != nil || prefixBytesRead < 5:
+func readEnvelope(dst *bytes.Buffer, src io.Reader, readMaxBytes int) (uint8, *Error) {
+	prefix := [5]byte{}
+	if _, err := io.ReadFull(src, prefix[:]); err != nil {
```

Comment:
Incorporates the fix from: https://github.com/connectrpc/connect-go/pull/580

Reply:
Looks like that contributor never got any feedback. Maybe it would be better to _accept_ (and merge) #580 (so the author gets credit for it in release notes) and then rebase this branch?

Reply:
Yep I agree, it's a good fix.